src/lib/zork/zork.ts

```diff
@@ -0,0 +1,378 @@
+// TODO: Replace console log calls with something more structured?
+// TODO: Do error states ever need to be indicated to clients rather than
+//       just logged locally?
+
+
+import * as constants from '../../generic_core/constants';
+import * as net from 'net';
+import * as node_server from '../socks/node/server';
+import * as node_socket from '../socks/node/socket';
+import * as socks_session from '../socks/session';
+import * as wrtc from 'wrtc';
+
+const RTC_PEER_CONFIG = {
+  iceServers: [
+    {url: 'stun:stun.l.google.com:19302'},
+    {url: 'stun:stun1.l.google.com:19302'},
+    {url: 'stun:stun.services.mozilla.com'}
+  ]
+};
+
+// The delimiter for Zork messages is just \n, but we use \r?\n for
+// MSG_DELIM_RE so that \r\n sent by e.g. a telnet client will also be treated
+// as a delimiter. https://en.wikipedia.org/wiki/Robustness_principle
+// "Be conservative in what you send, be liberal in what you accept."
+const MSG_DELIM_RE = /\r?\n/;
+
+const SOCKS_HOST = '0.0.0.0';
+const SOCKS_PORT = 9999;
+
+const ZORK_SERVER_PORT_START = 9000;
+const ZORK_SERVER_PORT_INCREMENT = 10;
+const ZORK_SERVER_BIND_MAXTRIES = 3;
+let zorkServerPort = ZORK_SERVER_PORT_START;
+let zorkServerBindNtries = 0;
+let nconnectionsMade = 0;
+let npeersGetting = 0;
+
+interface ParsedCmd {
+  id: string;        // e.g. 'ping', 'give', 'get', 'transform', etc.
+  source: string;    // e.g. 'transform with caesar'
+  tokens: string[];  // e.g. ['transform', 'with', 'caesar']
+}
+
+interface Context {
+  clientId: number;
+  socket: net.Socket;
+  mode: string;  // 'give', 'get', or null (still awaiting a 'give' or 'get' command)
+  peerConn: any; // using wrtc.RTCPeerConnection upsets tsc
+  transformer: any;
+}
+
+
+// Command handlers:
+
+const handleCmdInvalid = (ctx: Context, cmd: ParsedCmd) => {
+  ctx.socket.write(`I don't understand that command. (${cmd.id})\n`);
+};
+
+const handleCmdPing = (ctx: Context) => {
+  ctx.socket.write(`ping\n`);
+};
+
+const handleCmdXyzzy = (ctx: Context) => {
+  ctx.socket.write(`Nothing happens.\n`);
+};
+
+const handleCmdVersion = (ctx: Context) => {
+  ctx.socket.write(`${constants.MESSAGE_VERSION}\n`);
+};
+
+const handleCmdQuit = (ctx: Context) => {
+  ctx.socket.end();
+};
+
+const handleCmdGetters = (ctx: Context) => {
+  ctx.socket.write(`${npeersGetting}\n`);
+};
+
+const handleCmdTransform = (ctx: Context, cmd: ParsedCmd) => {
+  const t1 = cmd.tokens[1];
+  if (t1 === 'with') {
+    ctx.transformer = {name: cmd.tokens[2]};
+  } else if (t1 === 'config') {
+    const idx = cmd.source.indexOf(' config ') + ' config '.length;
+    const config = cmd.source.substring(idx);
+    ctx.transformer = {config: config};
+  } else {
+    ctx.socket.write(`usage: transform (with name|config json)\n`);
+  }
+};
+
+/*
+ * Start giving access to the peer.
+ * Create a local socks server and connect it to the peer connection.
+ */
+const handleCmdGive = (ctx: Context) => {
```

Comment:
By the way, I think when we were sketching this out in person we had the give and get command handlers flipped after all, so I switched them so that they're set properly in this PR - hopefully!